packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ClientProvider.cs

```diff
@@ -31,18 +63,245 @@ public ClientProvider(InputClient inputClient)
+        /// <param name="primaryCtorOrderedParams">The ordered parameters for the primary constructor.</param>
+        /// <param name="endpointParam">The endpoint parameter for the client.</param>
+        private ConstructorProvider BuildSecondaryConstructor
+            (List<ParameterProvider> requiredParams,
+            List<ParameterProvider> primaryCtorOrderedParams,
+            ParameterProvider? endpointParam)
+        {
+            var isEndpointRequired = endpointParam != null && endpointParam.DefaultValue == null;
+            List<ParameterProvider> secondaryCtorParameters = isEndpointRequired ? [endpointParam, .. requiredParams] : requiredParams;
+
+            // initialize the arguments for the primary constructor
+            var primaryCtorArgs = new List<ValueExpression>(primaryCtorOrderedParams.Count);
+            foreach (var p in primaryCtorOrderedParams)
+            {
+                if (p.DefaultValue == null && p.Validation != ParameterValidationType.None)
+                {
+                    primaryCtorArgs.Add(p);
+                }
+                else
+                {
+                    primaryCtorArgs.Add(p.DefaultValue ?? DefaultOf(p.Type));
+                }
+            }
+
+            var primaryCtorInitializer = new ConstructorInitializer(false, primaryCtorArgs);
+            var constructorSignature = new ConstructorSignature(
+                Type,
+                _publicCtorDescription,
+                MethodSignatureModifiers.Public,
+                secondaryCtorParameters,
+                Initializer: primaryCtorInitializer);
+
+            return new ConstructorProvider(
+                constructorSignature,
+                MethodBodyStatement.Empty,
+                this);
+        }
+
+        private List<ParameterProvider> GetSubClientCtorParams()
+        {
+            var orderedParams = new List<ParameterProvider> { PipelineField.AsParameter };
+            ParameterProvider? endpointParam = null;
+
+            // add the auth parameter
+            if (_apiKeyAuthParameter != null)
+            {
+                orderedParams.Add(_apiKeyAuthParameter);
+            }
+
+            // add the client parameters, except the endpoint parameter
+            foreach (var parameter in _clientParameters)
+            {
+                if (parameter.Name == KnownParameters.Endpoint.Name)
+                {
+                    endpointParam = parameter;
+                }
+                else
+                {
+                    orderedParams.Add(parameter);
+                }
+            }
+
+            // order the constructor parameters as (Pipeline, Auth, ClientParams, Endpoint).
+            if (endpointParam != null)
+            {
+                orderedParams.Add(endpointParam);
+            }
+
+            return orderedParams;
+        }
+
+        private ConstructorProvider BuildMockingConstructor()
```

Comment:
Can this be a static method on ConstructorProvider?  It seems any type provider might want to add this?